### PR TITLE
[JENKINS-27315] Add environment variable for git commit author email

### PR DESCRIPTION
**Pull request by damienbiggs**
Surprised no one had added this before. If someone could show me an alternative
way  to get the commit's author's email, that would be great.
This checks the commit's text for the author line and parses out the email
address.
The environment variable is set as GIT_COMMIT_AUTHOR_EMAIL.
My team use this to email the person who comitted a piece of code at the end of
a  build.

### CODE REVIEW

**Comment by KostyaSha** [comment]
`Surprised ` `My team use this to email the person who comitted a piece of code at the end of
a build.` email-ext support this out of the box, this is called "developers" that calculate emails from changed commits. Also build can be caused by multiple commits with different emails.

File: src/main/java/hudson/plugins/git/util/Build.java [hunk 1]
```diff
@@ -47,20 +47,23 @@
      */
     public Revision revision;
 
+    public String revisionAuthor;
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/git/util/Build.java]: imho this info shouldn't exist here.

File: src/main/java/hudson/plugins/git/util/BuildData.java [hunk 2]
```diff
@@ -157,6 +157,16 @@ public Build getLastBuildOfBranch(String branch) {
...
 
+    /**
+     * Gets revision author of the previous build.
+     * @return revision author of the last build.
+     *    May be null will be returned if nothing has been checked out (e.g. due to wrong repository or branch)
+     */
+    @Exported
+    public @CheckForNull String getLastBuiltRevisionAuthor() {
+        return lastBuild==null?null:lastBuild.revisionAuthor;
+    }
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/git/util/BuildData.java]: BuildData is designed for tracking hash+branch name and not for storing unrelated info.

**Comment by MarkEWaite** [comment]
I'd prefer that you use the email-ext plugin for this . It has much more capability and integrates nicely into the Jenkins user interface.

Can you confirm that email-ext meets your needs?

**Comment by damienbiggs** [comment]
I tested using email-ext and it's not quite what I need.
It's a fine plugin but adds a bit more configuration to each job to configure similar to the existing Jenkins email plugin.

It seems that I need to add multiple triggers to generate build is back to stable, unstable, failure emails etc.

I can add a Failure (any) trigger but it loses some contextual information.
On top of that the default format of the messages is different to the Jenkins emails, not a major issue but it would break any existing email filters that people have.

If people don't see enough general value in the commit author email being exposed, that's fine.
I can just use a locally built plugin on our Jenkins.

File: src/main/java/hudson/plugins/git/util/BuildData.java (hunk 2 above)
Comment (damienbiggs) [inline src/main/java/hudson/plugins/git/util/BuildData.java]: This definition of BuidData is arbitrary imo. If you're tracking a hash you're also indirectly tracking the author associated with that has and any other commit metadata that's deemed appropriate.

**Comment by damienbiggs** [comment]
Forgot to mention about a build having multiple commits.
This is actually one of the reasons that I want to get the latest commit's author (similar to what the GIT_COMMIT environmental variable represents today)

This might be specific to our usage but we have several pre commit jobs that developers can run.
These are based off their private branches.
So if a developer hasn't run a precommit job in a while, their next build might include multiple commits that have been pushed to master in the mean time.

In this (admittedly specific) circumstance, I don't want all the authors of the intervening commits to be emailed on a build failure. I only want to email the latest commit author.

**Comment by KostyaSha** [comment]
email-ext cover all cases for emailing, default jenkins emailer absolutely lacks on feature. You can also get email author by using shell execution and injecting any generated property file content into job variables with 'env-inject' plugin, or even wrote groovy code. 
imho this PR looks like hack for single use case.

**Comment by damienbiggs** [comment]
I realize that you can use a combination of execute shell and env inject to create the environment variable. However for a jenkins setup where you need to do this say 15 to 20 times, it's unnecessary noise. Imo it's much cleaner and less hacky to rely on the git plugin to provide environment variables such as this one.

Also as I mentioned, the email-ext plugin (I'm open to correction) will email everyone who pushed code to the repository since the last build. At times this is not desired behavior. And whilst the default jenkins emailer indeed has less features it has more sensible default configuration out of the box.

**Comment by MarkEWaite** [comment]
@damienbiggs I appreciate your desire for simplicity in your use case, but I'm unwilling to accept this change into the git plugin.

The use case is too narrow and there are reasonable ways to accomplish the same thing, without adding a new variable, and particularly without increasing the amount of information stored in BuildData.  I'm closing this commit.

Thanks very much for your submission, but this doesn't feel like the right thing for the git plugin.

**Comment by damienbiggs** [comment]
That's fine, I can still use this version on our jenkins.
Thanks for the response @MarkEWaite 

@KostyaSha you should mollify your comments on pull requests.
On both this pull request and others, you come across as rude and abrasive.

**Comment by KostyaSha** [comment]
@damienbiggs do good code and i will raise :+1:

**Comment by MarkEWaite** [comment]
@KostyaSha, I think you misunderstood (or ignored) the intent of the comment from @damienbiggs .  

As far as I can tell, the Jenkins project seeks to actively encourage contributors at various levels. I first started contributing by answering questions on the mailing list.  After some time of positive interactions on the mailing list, I started writing automated tests for the git plugin and the git client plugin.  Those were submitted by pull requests.  Eventually, the plugin owners decided that it would be reasonable to grant me commit access.  After that, eventually I became confident enough to review changes, and release new versions of the plugin.  It was an evolutionary process for me, and harsh comments along the way probably would have caused me to be much less willing to help.

Many contributors may be making a first attempt to help the project, and need tutoring and encouragement rather than feedback which they may perceive as too direct or too critical.

The interactions in the Jenkins community are much more "contributor-friendly" than (for example) the OpenBSD community, or even the Linux kernel community.  I think Damien is trying to encourage you to keep the Jenkins culture, rather than pushing towards the OpenBSD culture.

**Comment by KostyaSha** [comment]
@MarkEWaite first of all i see no problems in my comments to this PR (except attempt to hack plugin for own use case, instead of using right tools). Secondary this is a standard open-source. I will be glad to see the same culture as OpenBSD/Linux, if jenkins had the same quality. In this PR i see pure user-experience with jenkins and email-ext plugin and starting requests with words "surprised me that you still have no X" - is a bad start for communicating in jenkins community. Jenkins has many reasons for having or not having something.
Third, this PR violates https://wiki.jenkins-ci.org/display/JENKINS/Extend+Jenkins docs. Btw, you can add this link in CONTRIBUTING.md

**Pull request closed without merging**